group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -524,37 +495,39 @@ public CompletableFuture<ListGroupsResponseData> listGroups(
             );
         }
 
-        final CompletableFuture<ListGroupsResponseData> future = new CompletableFuture<>();
```

Comment:
so instead of having one future and the results in a list, we now have a list of futures where each list item is a list of the groups per topic partition?

Reply:
The end result is still a future containing a `ListGroupsResponseData` response. However, instead of building directly and having to lock, I rely more on the future operations. As you said, I first build a list of future where each contain a list of groups for a __consumer_offsets partition, then I combine them into the final future.